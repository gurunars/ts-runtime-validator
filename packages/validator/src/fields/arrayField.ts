import { Field, withErrorDecoration } from '../core'
import { Json } from '../Json'

const FieldSymbol = Symbol('@validator/fields.ArrayField')

class ArrayField<T> implements Field<T[]> {
  constructor(private readonly itemField: Field<T>) {}
  type = FieldSymbol

  validate(value: any): T[] {
    if (!Array.isArray(value)) {
      throw 'Not an array'
    }
    return value.map(
      (it, index) => withErrorDecoration(index, () => this.itemField.validate(it))
    )
  }
  serialize(deserialized: T[]): Json {
    return deserialized.map(
      (it, index) => withErrorDecoration(index, () => this.itemField.serialize(it) as unknown as Json)
    )
  }
  getParams(): Json {
    return {
      itemSpec: this.itemField.getParams(),
    }
  }
}

const arrayField = <T> (
  itemField: Field<T>,
): ArrayField<T> =>
    new ArrayField(itemField)

arrayField.type = FieldSymbol

export default arrayField
