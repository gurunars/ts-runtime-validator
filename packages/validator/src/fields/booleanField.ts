import { Field } from '../core'
import { Json } from '../Json'
import { Any } from '../util-types'
import { WithRegExp, WithStringInputSupport } from '../WithStringInputSupport'

const FieldSymbol = Symbol('@validator/fields.BooleanField')

class BooleanField implements Field<boolean>, WithStringInputSupport {
  type = FieldSymbol

  getFieldWithRegExp(): Field<Any> & WithRegExp {
    return new BooleanFieldWithRegExp()
  }
  validate(value: any): boolean {
    if (value !== true && value !== false) {
      throw 'Not a boolean'
    }
    return value
  }
  serialize(deserialized: boolean): Json {
    return deserialized
  }
}

class BooleanFieldWithRegExp extends BooleanField implements WithRegExp {

  regex() {
    return /true|false|1|0/
  }

  validate(value: any): boolean {
    if (value === 'true' || value === '1') {
      value = true
    }
    if (value === 'false' || value === '0') {
      value = false
    }
    return super.validate(value)
  }

}

const booleanField = (): BooleanField => new BooleanField()

booleanField.type = FieldSymbol

export default booleanField
