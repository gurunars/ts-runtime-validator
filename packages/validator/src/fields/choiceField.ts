import { Field } from '../core'
import { WithRegExp, WithStringInputSupport } from '../WithStringInputSupport'
import { escapeRegex } from '../utils'
import { Primitive, Json } from '../Json'

type Params<Choices> = {
  readonly choices: Choices
}

class ChoiceField<
  Choice extends Primitive,
> implements Field<Choice>, WithStringInputSupport {
  private choicesSet: Set<Primitive>

  constructor(readonly params: Params<readonly Choice[]>) {
    this.choicesSet = new Set(params.choices)
  }
  getFieldWithRegExp(): Field<unknown> & WithRegExp {
    return new ChoiceFieldWithRegExp(this.params)
  }

  validate(value: any): Choice {
    if (this.choicesSet.has(value)) {
      return value as Choice
    }
    throw 'Invalid choice'
  }
  serialize(deserialized: Choice): Json {
    return deserialized as unknown as Primitive
  }
  getParams() {
    return this.params
  }

}

class ChoiceFieldWithRegExp<
  Choice extends Primitive
> extends ChoiceField<Choice> implements WithRegExp {

  private fullChoiceMap: Map<any, Primitive>

  constructor(readonly params: Params<readonly Choice[]>) {
    super(params)
    this.fullChoiceMap = new Map<any, Primitive>()

    params.choices.forEach(it => {
      this.fullChoiceMap.set(it, it)
      this.fullChoiceMap.set(it.toString(), it)
    })
  }

  regex() {
    return new RegExp(Object.keys(this.params.choices)
      .map(it => it.toString())
      .map(escapeRegex)
      .join('|')
    )
  }

  validate(value: any): Choice {
    return super.validate(this.fullChoiceMap.get(value))
  }

}

const choiceField = <
  Choice extends Primitive,
> (...choices: readonly Choice[]): ChoiceField<Choice> =>
    new ChoiceField({
      choices,
    })

export default choiceField

