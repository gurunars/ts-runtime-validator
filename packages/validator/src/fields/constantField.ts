import { Field } from '../core'
import { Primitive } from '../Json'
import { field } from '../registry'

export interface ConstantField<Constant extends Primitive> extends Field<Constant> {
  readonly constant: Constant
}

export default field('@validator/fields.ConstantField', <Constant extends Primitive> (
  constant: Constant
): ConstantField<Constant> => ({
    constant,
    validate: (value: any): Constant => {
      if (value !== constant) {
        throw 'Constant does not match the requirement'
      }
      return constant
    },
    serialize: () => constant
  }))
