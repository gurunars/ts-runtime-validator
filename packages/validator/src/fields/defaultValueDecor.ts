import { Field, Json } from '../core';
import { merge } from '../utils';

class DefaultFieldDecorator<T> implements Field<T> {
  private innerField: Field<T>;
  private defaultValue: T;

  constructor(innerField: Field<T>, defaultValue: T) {
    this.innerField = innerField;
    this.defaultValue = defaultValue;
  }

  validate(value: any): T {
    if (value === undefined) {
      value = this.defaultValue
    }
    return this.innerField.validate(value)
  }
  serialize(deserialized: T): Json {
    return this.innerField.serialize(deserialized)
  }
  getParams() {
    return merge({
      defaultValue: this.defaultValue
    }, this.innerField.getParams())
  }
}

const defaultValueDecor = <T> (innerField: Field<T>, defaultValue: T): Field<T> =>
  new DefaultFieldDecorator(innerField, defaultValue);

export default defaultValueDecor;
