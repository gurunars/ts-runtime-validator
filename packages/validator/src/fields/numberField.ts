import { Field } from '../core'
import { Json } from '../Json'
import { WithRegExp, WithStringInputSupport } from '../WithStringInputSupport'

type Params = {
  canBeFloat?: boolean
}

class NumberField implements Field<number>, WithStringInputSupport {
  protected params?: Params

  constructor(params?: Params) {
    this.params = params
  }
  getFieldWithRegExp(): Field<unknown> & WithRegExp {
    return new NumberFieldWithRegExp(this.params)
  }

  validate(value: any): number {
    if (typeof value !== 'number') {
      throw 'Not a number'
    }
    if (!this.params?.canBeFloat && value !== Math.floor(value)) {
      throw 'Not an int'
    }
    return value
  }
  serialize(deserialized: number): Json {
    return deserialized
  }
  getParams() {
    return {
      ...this.params,
    }
  }
}

class NumberFieldWithRegExp extends NumberField implements WithRegExp {

  regex() {
    const parts: string[] = []
    parts.push('-?')
    parts.push('\\d+')
    if (this.params?.canBeFloat) {
      parts.push('(\\.\\d+)?')
    }

    return RegExp(parts.join(''))
  }

  validate(value: any) {
    return super.validate(Number.parseFloat(value))
  }

}

const numberField = (params?: Params): NumberField => new NumberField(params)

export default numberField
