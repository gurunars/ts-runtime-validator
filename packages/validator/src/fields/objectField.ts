import {
  Field,
  ValidatorSpec,
  validate,
  serialize,
} from '../core'
import { declareField } from '../docRegistry'
import { Json } from '../Json'
import { Any } from '../util-types'

class ObjectField<DeserializedType extends Record<string, Any>> implements Field<DeserializedType> {
  constructor(readonly objectSpec: ValidatorSpec<DeserializedType>) {}

  validate(value: any): DeserializedType {
    if (typeof value !== 'object' || value === null) {
      throw 'Not an object'
    }
    return validate(this.objectSpec, value) as DeserializedType
  }
  serialize(deserialized: DeserializedType): Json {
    return serialize(this.objectSpec, deserialized as any)
  }
}

export default declareField('@validator/fields.ObjectField', ObjectField) as
<DeserializedType extends Record<string, Any>> (
  objectSpec: ValidatorSpec<DeserializedType>,
) => ObjectField<DeserializedType>
