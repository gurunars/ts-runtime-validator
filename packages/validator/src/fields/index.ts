import arrayField from './arrayField'
import booleanField from './booleanField'
import numberField from './numberField'
import objectField from './objectField'
import stringField from './stringField'
import choiceField from './choiceField'
import withDefault from './withDefault'
import optional from './optional'
import unionField from './unionField'
import constantField from './constantField'
import $ from './segmentField'
import wildcardObjectField from './wildcardObjectField'
import undefinedField from './undefinedField'

export {
  arrayField,
  booleanField,
  numberField,
  objectField,
  stringField,
  choiceField,
  withDefault,
  optional,
  unionField,
  constantField,
  $,
  wildcardObjectField,
  undefinedField,
}
