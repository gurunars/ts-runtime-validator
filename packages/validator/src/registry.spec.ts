import { numberField, stringField, objectField, booleanField } from './fields'

import createRegistry, { $, Registry } from './registry'

const normalRegistryPairs: Registry = [
  $(stringField, (field) => ({
    regexp: field.regex.source,
    type: 'string'
  })),
  $(numberField, (field) => ({
    canBeFloat: field.params?.canBeFloat || false,
    type: 'number'
  })),
  $(objectField, (field, requestRepresentation) => Object.fromEntries(
    Object.entries(field.objectSpec).map(
      ([key, subfield]) => [key, requestRepresentation(subfield)]
    )
  ))
]

describe('createRegistry', () => {
  it('correctly returns a representation of a schema with known field', () => {
    const registry = createRegistry(normalRegistryPairs)
    const schema = objectField({
      title: stringField(),
      page: objectField({
        subtitle: stringField(),
        pageCount: numberField()
      })
    })
    expect(registry(schema)).toMatchSnapshot()
  })

  it('throws an error if a duplicate type is registered', () => {
    expect(() => createRegistry([...normalRegistryPairs, $(stringField, () => undefined)]))
      .toThrow(`Found duplicate of type declaration '${stringField.type}'`)
  })

  it('throws an error if field type is unknown', () => {
    const registry = createRegistry(normalRegistryPairs)
    const schema = objectField({
      flag: booleanField()
    })
    expect(() => registry(schema))
      .toThrow(`Could not find field of type '${booleanField.type}'`)
  })
})
