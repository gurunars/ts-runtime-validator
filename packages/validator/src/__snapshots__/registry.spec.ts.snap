// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createRegistry correctly returns a representation of a schema with known field 1`] = `
Object {
  "page": Object {
    "pageCount": Object {
      "canBeFloat": false,
      "type": "number",
    },
    "subtitle": Object {
      "regexp": undefined,
      "type": "string",
    },
  },
  "title": Object {
    "regexp": undefined,
    "type": "string",
  },
}
`;
